utils/relay-manager.js

```diff
@@ -1,26 +1,49 @@
+/* global artifacts */
+const BN = require("bn.js");
 const ethers = require("ethers");
 const { ETH_TOKEN } = require("./utilities.js");
 const utils = require("./utilities.js");
 
+const ITokenPriceRegistry = artifacts.require("ITokenPriceRegistry");
+const IGuardianStorage = artifacts.require("IGuardianStorage");
+const IFeature = artifacts.require("IFeature");
+
 class RelayManager {
-  setRelayerManager(relayerManager) {
+  async setRelayerManager(relayerManager) {
     this.relayerManager = relayerManager;
+
+    const guardianStorageAddress = await relayerManager.guardianStorage();
+    this.guardianStorage = await IGuardianStorage.at(guardianStorageAddress);
   }
 
+  // Relays without refund by default, unless the gasPrice is explicitely set to be >0
   async relay(_module, _method, _params, _wallet, _signers,
-    _relayerAccount,
-    _estimate = false,
-    _gasLimit = 2000000,
-    _nonce,
     _gasPrice = 0,
     _refundToken = ETH_TOKEN,
-    _refundAddress = ethers.constants.AddressZero,
-    _gasLimitRelay) {
-    if (!_gasLimitRelay) { _gasLimitRelay = (_gasLimit * 1.1); } // eslint-disable-line no-param-reassign
-    const relayerAccount = _relayerAccount || await utils.getAccount(9);
-    const nonce = _nonce || await utils.getNonceForRelay();
-    const methodData = _module.contract.methods[_method](..._params).encodeABI();
+    _refundAddress = ethers.constants.AddressZero) {
+    const relayerAccount = await utils.getAccount(9);
+    const nonce = await utils.getNonceForRelay();
     const chainId = await utils.getChainId();
+    const methodData = _module.contract.methods[_method](..._params).encodeABI();
+
+    const gasLimit = await this.getGasLimitRefund(_module, _method, _params, _wallet, _signers, _gasPrice);
+    // Uncomment when debugging gas limits
+    // await this.debugGasLimits(_module, _method, _params, _wallet, _signers);
+
+    // When the refund is in token (not ETH), calculate the amount needed for refund
+    let gasPrice = _gasPrice;
+    if (_refundToken !== ETH_TOKEN) {
+      const tokenPriceRegistryAddress = await this.relayerManager.tokenPriceRegistry();
+      const tokenPriceRegistry = await ITokenPriceRegistry.at(tokenPriceRegistryAddress);
+      const tokenPrice = await tokenPriceRegistry.getTokenPrice(_refundToken);
+
+      // How many tokens to pay per unit of gas spent
+      // refundCost = gasLimit * gasPrice
+      // tokenAmount = refundCost * 10^18 / tokenPrice
+      const tokenAmount = new BN(10).pow(new BN(18)).muln(_gasPrice).div(tokenPrice);
```

Comment:
This "tokenAmount" variable is in fact the gasPrice so maybe we could instead write 
```
gasPrice = new BN(10).pow(new BN(18)).muln(_gasPrice).div(tokenPrice).toNumber();
```